crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,261 @@
+/// Marker component added on the main world entities that are synced to the Render World in order to keep track of the corresponding render world entity
+pub struct RenderEntity(Entity);
+impl RenderEntity {
+    #[inline]
+    pub fn id(&self) -> Entity {
+        self.0
+    }
+}
+
+#[derive(Component, Deref, Clone, Debug)]
+/// Marker component added on the render world entities to keep track of the corresponding main world entity
+pub struct MainEntity(Entity);
+impl MainEntity {
+    #[inline]
+    pub fn id(&self) -> Entity {
+        self.0
+    }
+}
+
+// marker component that indicates that its entity needs to be despawned at the end of every frame.
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[component(storage = "SparseSet")]
+pub struct TemporaryRenderEntity;
+
+pub(crate) enum EntityRecord {
+    // When an entity is spawned on the main world, notify the render world so that it can spawn a corresponding entity. This contains the main world entity
+    Added(Entity),
+    // When an entity is despawned on the main world, notify the render world so that the corresponding entity can be despawned. This contains the render world entity.
+    Removed(Entity),
+}
+
+// Entity Record in MainWorld pending to Sync
+#[derive(Resource, Default, Deref, DerefMut)]
+pub(crate) struct PendingSyncEntity {
+    records: Vec<EntityRecord>,
+}
+
+pub(crate) fn entity_sync_system(main_world: &mut World, render_world: &mut World) {
+    main_world.resource_scope(|world, mut pending: Mut<PendingSyncEntity>| {
+        // TODO : batching record
+        for record in pending.drain(..) {
+            match record {
+                EntityRecord::Added(e) => {
+                    if let Some(mut entity) = world.get_entity_mut(e) {
+                        match entity.entry::<RenderEntity>() {
+                            bevy_ecs::world::Entry::Occupied(_) => {
+                                warn!("Attempting to synchronize an entity that has already been synchronized!");
+                            }
+                            bevy_ecs::world::Entry::Vacant(entry) => {
+                                let id = render_world.spawn(MainEntity(e)).id();
+
+                                entry.insert(RenderEntity(id));
+                            }
+                        };
+                    }
+                }
+                EntityRecord::Removed(e) => {
+                    if let Some(ec) = render_world.get_entity_mut(e) {
+                        ec.despawn_recursive();
+                    };
+                }
+            }
+        }
+    });
+}
+
+// TODO: directly remove matched archetype for performance
+pub(crate) fn despawn_temporary_render_entity(
+    world: &mut World,
+    state: &mut SystemState<Query<Entity, With<TemporaryRenderEntity>>>,
+    mut local: Local<Vec<Entity>>,
+) {
+    let query = state.get(world);
+
+    local.extend(query.iter());
+
+    // ensure next frame allocation keeps order
```

Comment:
```suggestion
    // Ensure next frame allocation keeps order
```

Also, can you expand the comment a bit more? What does this mean, and why are we doing it?